Fix for issue 1280:  Notification of Newer Opencast Version in Admin UI

mheyen opened the pull request:
This patch adds a notification in the admin UI (behind the bell) if there is a newer version of Opencast available.
The notification shows:
- if there is a new minor release (e.g. 8.4 -> 8.5)
- if there is a new major release (e.g. 7.4 -> 8.1)
- if your version is no longer supported (older than two major releases)

Added service in restServiceMonitor.js :
- compares your version to the latest version and adds notification accordingly.
- adds link in the notification to https://docs.opencast.org/r/<LATEST VERSION>/admin/ to enable the user to update if neccesary.

Added new endpoint at VersionEndpoint.java :
- retrieves latest version from https://api.github.com/repos/opencast/opencast/releases/latest .
- updates every hour.
- provides version via /admin-ng/oc-version/version.json .

This fixes #1280

### Your pull request should…

* [ ] have a concise title
* [ ] [close an accompanying issue](https://help.github.com/en/articles/closing-issues-using-keywords) if one exists
* [ ] be against the correct branch (features can only go into develop)
* [ ] include migration scripts and documentation, if appropriate
* [ ] pass automated testing
* [ ] have a clean commit history
* [ ] have proper commit messages (title and body) for all commits
* [ ] have appropriate tags applied

mheyen committed:
Fix for issue 1280:  Notification of Newer Opencast Version in Admin UI

This patch adds a notification in the admin UI behind the bell, if there is a newer version of Opencast available.
The notification shows:
- if there is a new minor release (e.g. 8.4 -> 8.5)
- if there is a new major release (e.g. 7.4 -> 8.1)
- if your version is no longer supported (older than two major releases)

Added service in restServiceMonitor.js :
- compares your version to the latest version and adds notification accordingly.
- adds link in the notification to https://docs.opencast.org/r/<LATEST VERSION>/admin/ to enable the user to update if neccesary.

Added new endpoint at VersionEndpoint.java :
- retrieves latest version from https://api.github.com/repos/opencast/opencast/releases/latest .
- updates every hour.
- provides version via /admin-ng/oc-version/version.json .

This fixes #1280

lkiesow commented on modules/admin-ui-frontend/app/scripts/shared/services/restServiceMonitor.js:
```diff
@@ -45,6 +48,42 @@ angular.module('adminNg.services')
+          } else if (my_version[0] == latest_version[0]) {
+            services.service[LATEST_VERSION_NAME].status = 'There is a minor update available.';
+            services.service[LATEST_VERSION_NAME].error = true;
+          } else if (parseInt(latest_version[0]) - parseInt(my_version[0]) <= 2) {
```
```suggestion
          } else if (parseInt(latest_version[0]) - parseInt(my_version[0]) < 2) {
```

Currently, 8.6 is the latest version and 8.x and 7.x is supported.
If someone has a 6.4 then `8 - 6 = 2` which should yield the `…is no longer supported. Please update.`

lkiesow commented on modules/admin-ui-frontend/app/scripts/shared/services/restServiceMonitor.js:
```diff
@@ -45,6 +48,42 @@ angular.module('adminNg.services')
+          services.service[LATEST_VERSION_NAME].docs_url =
+            'https://docs.opencast.org/r/' + latest_version[0] + '.x/admin/';
+
+          if (parseFloat(my_version) >= parseFloat(latest_version) || my_version.endsWith('SNAPSHOT')) {
```
```suggestion
          if (parseFloat(my_version) >= parseFloat(latest_version)
             || (my_version[0] == latest_version[0] && my_version.endsWith('SNAPSHOT'))) {
```

This could also be `4.0.0-SNAPSHOT` which could be identified as very old.

lkiesow commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/VersionEndpoint.java:
```diff
@@ -0,0 +1,113 @@
+      version = (String) data.get("tag_name");
+      lastUpdated = System.currentTimeMillis() / 1000L;
+    } catch (Exception e) {
+      logger.error("Error while parsing the version: " + e.toString());
```
I would make this a warning since it's not something an admin would (or most likely even could) fix.
And use the logger's functionality to handle exceptions:

```suggestion
      logger.error("Error while parsing the Opencast version from GitHub", e);
```

mheyen replied:
How do I make this a warning? 
Does the logger.error() function handle the Exception in a way such that it will create a warning?

lkiesow replied:
`logger.warn(…)`

lkiesow reviewed (changes requested):
Mostly nitpicking…

mheyen committed:
Update modules/admin-ui-frontend/app/scripts/shared/services/restServiceMonitor.js

removed minor mistake in version control.

Co-authored-by: Lars Kiesow <lkiesow@uos.de>

mheyen committed:
Update modules/admin-ui-frontend/app/scripts/shared/services/restServiceMonitor.js

Older versions including 'SNAPSHOT' are now identified.

Co-authored-by: Lars Kiesow <lkiesow@uos.de>

mheyen committed:
Update modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/VersionEndpoint.java

changed error to warning message.

mheyen committed:
Merge remote-tracking branch 'origin/fix-1280' into fix-1280

gregorydlogan commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/VersionEndpoint.java:
```diff
@@ -0,0 +1,113 @@
+  @Produces(MediaType.APPLICATION_JSON)
+  @RestQuery(name = "latestversion", description = "Returns the latest Opencast version",
+          returnDescription = "Returns the latest Opencast version retrieved from GitHub",
+          reponses = { @RestResponse(responseCode = SC_OK, description = "The latest Opencast version.") })
```
This should be `re**s**ponses`, especially since I'm looking at #1766 too :)

mheyen replied:
That is absolutely correct ;)

gregorydlogan commented on modules/admin-ui-frontend/app/scripts/shared/services/restServiceMonitor.js:
```diff
@@ -58,11 +98,7 @@ angular.module('adminNg.services')
         services.service[AMQ_NAME].error = true;
       }
     }).catch(function(err) {
-      Monitoring.populateService(AMQ_NAME);
```
What was the reasoning behind removing this?

mheyen replied:
Monitoring.setError() already calls the Monitoring.populateService() function. 
Therefore I thought it would not be necessary to call it twice. 
Basically it was just to remove redundant code.

gregorydlogan commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/VersionEndpoint.java:
```diff
@@ -0,0 +1,113 @@
+          returnDescription = "Returns the latest Opencast version retrieved from GitHub",
+          reponses = { @RestResponse(responseCode = SC_OK, description = "The latest Opencast version.") })
+  public String getVersion() {
+    if (System.currentTimeMillis() / 1000L - lastUpdated >= 3600) {
```
Probably only need to check this once, either here, or in `updateVersion`

lkiesow replied:
It does make sense to check this twice. If you check it just in `updateVersion` you have basically a single-threaded web server for this endpoint since that function is synchronized. If you have it just here, you can have multiple updates triggered at once which could cause GitHub temporarily blocking you.

gregorydlogan replied:
Ah, I see what you're saying.  I'm not sure that avoiding the single thread matters, but you're correct in that this approach will scale better.

mheyen committed:
Update modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/VersionEndpoint.java

gregorydlogan commented on modules/admin-ui-frontend/app/scripts/shared/services/restServiceMonitor.js:
```diff
@@ -45,6 +48,43 @@ angular.module('adminNg.services')
+          var my_version = response_my_version.data.version;
+          var latest_version = response_latest_version.data;
+          services.service[LATEST_VERSION_NAME].docs_url =
+            'https://docs.opencast.org/r/' + latest_version[0] + '.x/admin/';
```
Testing this, I patched out the endpoint to return "10.0" as the version.  This line generates links to `https://docs.opencast.org/r/1.x/admin/` rather than `https://docs.opencast.org/r/10.x/admin/`

Otherwise looks good and works as expected!

mheyen replied:
You are right. I patched this and it should work now. Thanks!

mheyen committed:
Update restServiceMonitor.js and VersionEndpoint

restServiceMonitor.js can now handle versions >= 10.0

fixed a mistake in VersionEndpoint.java .
It now searches in the list of latest releases for the highest version.
previously it received only the latest release, which could have been a minor release for an older version.
